internal/linter/check_pk_type.go

```diff
@@ -0,0 +1,69 @@
+package linter
+
+import (
+	"fmt"
+	"regexp"
+	"strings"
+
+	"github.com/skeema/skeema/internal/tengo"
+)
+
+func init() {
+	rule := Rule{
+		CheckerFunc:     TableBinaryChecker(pkTypeChecker),
+		Name:            "pk-type",
+		Description:     "Only allow primary keys to have types listed in --allow-pk-type",
+		DefaultSeverity: SeverityIgnore,
+	}
+	rule.RelatedListOption(
+		"allow-pk-type",
+		"tinyint, smallint, mediumint, int, bigint, varbinary, binary, date, datetime, timestamp, time, year, decimal",
+		"List of allowed data types for --lint-pk-type",
+		false, // permit empty list
+	)
+	RegisterRule(rule)
+}
+
+func pkTypeChecker(table *tengo.Table, createStatement string, _ *tengo.Schema, opts Options) *Note {
+	if table.PrimaryKey == nil {
+		// This checker is expected to be used in combination with check-pk
+		// so if there is no PK, we don't need to complain twice.
+		return nil
+	}
+	allowedTypes := opts.AllowList("pk-type")
+	if len(allowedTypes) == 0 {
+		return nil // disable check if allow-pk-type is empty
+	}
+	allowedStr := strings.Join(allowedTypes, ", ")
+	cols := table.ColumnsByName()
+	for _, part := range table.PrimaryKey.Parts {
+		if col, ok := cols[part.ColumnName]; ok {
+			colType := baseColType(col.TypeInDB)
+			if !opts.IsAllowed("pk-type", colType) {
+				message := fmt.Sprintf(
+					"Column %s of table %s is using data type %s, which is not configured to be permitted. The following data types are listed in option allow-pk-type: %s.",
+					col.Name, table.Name, col.TypeInDB, allowedStr,
+				)
+				return &Note{
+					LineOffset: FindColumnLineOffset(col, createStatement),
+					Summary:    "Column data type not permitted for PRIMARY KEY",
+					Message:    message,
+				}
+			}
+		}
+	}
+	return nil
+}
+
+var reBaseType = regexp.MustCompile(`^([a-zA-Z]+)`)
+
+// baseColType normalizes the type to remove display width/length
+// and any options like UNSIGNED/ZEROFILL. It is more exhaustive than
+// the check in check_display_width.go.
+func baseColType(colType string) string {
```

Comment:
This is totally fine for now, but it gave me an idea for a future tweak: perhaps tengo.Column should get an additional field, mirroring the value of `information_schema.columns.data_type`? If that value already lines up with this conversion logic, it would simplify things here, as well as probably in a bunch of other similar code paths elsewhere

Reply:
Yeah, that makes sense to me: either a field or a method on the struct.

I thought about adding it, but went with this simplified implementation. Happy to change it if you prefer.